### ISSUE TITLE
tf.dataset + tf.estimator  slow, starving CPU/GPU

### ISSUE THREAD

**Issue by mathlf2015** [tensorflow/tensorflow#24960]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution  Linux centos
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 1.9.0
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: cudatoolkit   8.0    and cudnn 7.1.3  
- GPU model and memory:  Tesla M40

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
I used the gpu to train the wide deep model with tf.dataset and tf.estimator .In the case that I have opened 4 gpus with os.environ['CUDA_VISIBLE_DEVICES'] = "0,1,2,3", but only one gpu utilization is 15%, the others are 0.Is it because the network structure of the deep network part cannot be parallel. So you can only use one gpu. Is there any other way to speed up the training of the model?
**Describe the expected behavior**

**Comment by jvishnuvardhan** [comment]
@mathlf2015 Could you provide as many details as possible to find the root cause of the issue. It would be great if you can provide a small code to reproduce the issue. Thanks!

**Comment by mathlf2015** [comment]
i read data from a csv file with 15488184 rows  and 20 columns,some part of code:
`def input_fn(filenames, num_epochs, batch_size=1):
    def parse_csv(line):
        print('Parsing', filenames)
        columns = tf.decode_csv(line, record_defaults=CSV_COLUMN_DEFAULTS)
        features = dict(zip(CSV_COLUMNS, columns))
        labels = features.pop(LABEL_COLUMN)

        return features, labels

    # Extract lines from input files using the Dataset API.
    dataset = tf.data.TextLineDataset(filenames) # can pass one filename or filename list

    # multi-thread pre-process then prefetch
    dataset = dataset.map(parse_csv, num_parallel_calls=50)
    # dataset.map(lambda features, target: (process_features(features), target))

    # We call repeat after shuffling, rather than before, to prevent separate epochs from blending together.
    # Shuffle, repeat, and batch the examples.
    if FLAGS.task_type == "train":
        dataset = dataset.shuffle(500)
    dataset = dataset.repeat(num_epochs)
    dataset = dataset.batch(batch_size)
    dataset = dataset.prefetch(5)
    iterator = dataset.make_one_shot_iterator()
    features, labels = iterator.get_next()
    # features = process_features(features)

    return features, labels

and the config part ：
session_config = tf.ConfigProto(log_device_placement=True,
                                    device_count={'GPU': 0},
                                    inter_op_parallelism_threads=FLAGS.inter_op,
                                    intra_op_parallelism_threads=FLAGS.intra_op
                                    )
config = tf.estimator.RunConfig().replace(session_config=session_config,
            save_checkpoints_steps=FLAGS.log_steps, log_step_count_steps=FLAGS.log_steps, save_summary_steps=FLAGS.log_steps)`

**Comment by dev-strender** [comment]
Could you say why you set ```device_count={'GPU': 0}``` in session_config?

This method prevents the Tensorflow Graph from using GPU.
I think deleting that parameter and running the model would work.

**Comment by mathlf2015** [comment]
> Could you say why you set `device_count={'GPU': 0}` in session_config?
> 
> This method prevents the Tensorflow Graph from using GPU.
> I think deleting that parameter and running the model would work.

This configuration is to automatically assign gpu instead of using gpu. I tried deleting and the results didn't improve.

**Comment by Goorman** [comment]
@mathlf2015 I have rewritten a year ago my DL pipeline in keras and it fixed all cpu starvation problems, training speed improved 10-fold. So my impression is that tf.data has some serious performance related bugs inside of it.

So use Keras or PyTorch, lol.

**Comment by mathlf2015** [comment]
> @mathlf2015 I have rewritten a year ago my DL pipeline in keras and it fixed all cpu starvation problems, training speed improved 10-fold. So my impression is that tf.data has some serious performance related bugs inside of it.
> 
> So use Keras or PyTorch, lol.

@Goorman Thank you very much for your answer. I have two questions. First, when I open 4 gpus, only one is in use, and secondly, the gpu used is at 25% utilisation. The utilization rate is very low. Can you use multiple gpus when using wide deep, have you encountered both of these problems? I really want to know how you solved it. Thank you

**Comment by sdll** [comment]
Any updates on this?

**Comment by sanjoy** [comment]
For these kind of performance issues, it is most productive if you poke at the [TensorBoard Profile](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) to see if can spot any bottlenecks.

**Comment by guotong1988** [comment]
Any updates on this?

**Comment by sanatmpa1** [comment]
@mathlf2015,

Can you try updating Tensorflow to latest stable version i.e `2.6.0` and let us know if the issue persists? Thanks!
